File: python-module/src/hashmap/pair_sum_unsorted.py
from typing import List

from utils import util_functions as uf

# Read array from console
array = uf.read_int_array()

# Print original array to console
uf.print_array(array)

target = uf.read_int()


def pair_sum_unsorted_two_pass(nums: List[int],
                               tar: int) -> List[int]:
    num_map = {}
    # First pass: Populate the hash map with each number and its index.
    for i, num in enumerate(nums):
        num_map[num] = i
    # Second pass: Check for each number's complement in the hash map.
    for i, num in enumerate(nums):
        complement = tar - num
        if complement in num_map and num_map[complement] != i:
            return [i, num_map[complement]]
    return []


print("Two pass approach: ", pair_sum_unsorted_two_pass(array, target))

File: java-module/src/hashmaps/PairSumUnsorted.java
package hashmaps;

import java.util.Arrays;
import java.util.HashMap;

import static util.ArrayUtils.*;

public class PairSumUnsorted {

    public static void main(String[] args) {
        System.out.println("Populate array of integers");
        int[] array = readArrayFromConsole();
        System.out.println("Unsorted array");
        printArrayToConsole(array);
        System.out.println("Enter target: ");
        int target = readIntegerFromConsole();
        System.out.println(String.format("Target is: %s", target));
        System.out.println("Complementary pair: " +
                Arrays.toString(pairSumUnsortedTwoPass(array, target)));
    }

    private static int[] pairSumUnsortedTwoPass(int[] array, int target) {
        var numMap = new HashMap<Integer, Integer>();
        // First pass: Populate the hash map with each number and its index.
        for (int i = 0; i < array.length; i++) {
            numMap.put(array[i], i);
        }
        // Second pass: Check for each number's complement in the hash map
        for (int i = 0; i < array.length; i++) {
            int num = array[i];
            int complement = target - num;
            if (numMap.containsKey(complement) && numMap.get(complement) != i) {
                return new int []{i, numMap.get(complement)};
            }
        }
        return new int []{};
    }
}
